metric/distribution/exph/exph.go

```diff
@@ -0,0 +1,289 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT
+
+package exph
+
+import (
+	"fmt"
+	"log"
+	"math"
+	"slices"
+
+	"go.opentelemetry.io/collector/pdata/pmetric"
+
+	"github.com/aws/amazon-cloudwatch-agent/metric/distribution"
+)
+
+type ExpHistogramDistribution struct {
+	max             float64
+	min             float64
+	sampleCount     float64
+	sum             float64
+	scale           int32
+	positiveBuckets map[int]uint64 // map of bucket index to count
+	negativeBuckets map[int]uint64 // map of bucket index to count
+	zeroThreshold   float64
+	zeroCount       uint64
+	unit            string
+}
+
+func NewExpHistogramDistribution() *ExpHistogramDistribution {
+	return &ExpHistogramDistribution{
+		max:             -math.MaxFloat64,
+		min:             math.MaxFloat64,
+		sampleCount:     0,
+		sum:             0,
+		scale:           0,
+		unit:            "",
+		positiveBuckets: map[int]uint64{},
+		negativeBuckets: map[int]uint64{},
+		zeroThreshold:   0,
+		zeroCount:       0,
+	}
+}
+
+func (d *ExpHistogramDistribution) Maximum() float64 {
+	return d.max
+}
+
+func (d *ExpHistogramDistribution) Minimum() float64 {
+	return d.min
+}
+
+func (d *ExpHistogramDistribution) SampleCount() float64 {
+	return d.sampleCount
+}
+
+func (d *ExpHistogramDistribution) Sum() float64 {
+	return d.sum
+}
+
+func (d *ExpHistogramDistribution) Unit() string {
+	return d.unit
+}
+
+func (d *ExpHistogramDistribution) Size() int {
+	size := len(d.negativeBuckets) + len(d.positiveBuckets)
+	if d.zeroCount > 0 {
+		size++
+	}
+	return size
+}
+
+// ValuesAndCounts outputs two arrays representing the midpoints of each exponential histogram bucket and the
+// counter of datapoints within the corresponding exponential histogram buckets
+func (d *ExpHistogramDistribution) ValuesAndCounts() ([]float64, []float64) {
+	values := []float64{}
+	counts := []float64{}
+
+	// iterate through positive buckets in descending order
+	posOffsetIndicies := make([]int, 0, len(d.positiveBuckets))
+	for offsetIndex := range d.positiveBuckets {
+		posOffsetIndicies = append(posOffsetIndicies, offsetIndex)
+	}
+	slices.Sort(posOffsetIndicies)
+	slices.Reverse(posOffsetIndicies)
```

Comment:
Instead of sorting and then reversing, we can use `maps.Keys` which returns an iterator and a sort function to reverse it while doing the sort.

https://pkg.go.dev/maps#Keys
https://pkg.go.dev/slices#SortedFunc

```suggestion
	posOffsetIndicies := slices.SortedFunc(maps.Keys(d.positiveBuckets), func(a, b int) int {
	  return cmp.Compare(b, a)
	})
```

Reply:
very nice, I'll update